text/0078-merkleized-metadata.md

```diff
@@ -0,0 +1,394 @@
+
+The chunking may not be the optimal case for every kind of offline wallet.
+
+## Testing, Security, and Privacy
+
+All implementations are required to strictly follow the RFC to generate the metadata hash. This includes which hash function to use and how to construct the metadata types tree. So, all implementations are following the same security criteria. As the chains will calculate the metadata hash at compile time, the build process needs to be trusted. However, this is already a solved problem in the Polkadot ecosystem by using reproducible builds. So, anyone can rebuild a chain runtime to ensure that a proposal is actually containing the changes as advertised.
+
+Implementations can also be tested easily against each other by taking some metadata and ensuring that they all come to the same metadata hash.
+
+Privacy of users should also not be impacted. This assumes that wallets will generate the metadata hash locally and don't leak any information to third party services about which chunks a user will send to their offline wallet. Besides that, there is no leak of private information as getting the raw metadata from the chain is an operation that is done by almost everyone.
+
+## Performance, Ergonomics, and Compatibility
+
+### Performance
+
+There should be no measurable impact on performance to Polkadot or any other chain using this feature. The metadata root hash is calculated at compile time and at runtime it is optionally used when checking the signature of a transaction. This means that at runtime no performance heavy operations are done. 
+
+###  Ergonomics & Compatibility
+
+The proposal alters the way a transaction is built, signed, and verified. So, this imposes some required changes to any kind of developer who wants to construct transactions for Polkadot or any chain using this feature. As the developer can pass `0` for disabling the verification of the metadata root hash, it can be easily ignored.
+
+## Prior Art and References
+
+[RFC 46](https://github.com/polkadot-fellows/RFCs/pull/46) produced by the Alzymologist team is a previous work reference that goes in this direction as well.
+
+On other ecosystems, there are other solutions to the problem of trusted signing. Cosmos for example has a standardized way of transforming a transaction into some textual representation and this textual representation is included in the signed data. Basically achieving the same as what the RFC proposes, but it requires that for every transaction applied in a block, every node in the network always has to generate this textual representation to ensure the transaction signature is valid.
+
+## Unresolved Questions
+
+None.
+
+## Future Directions and Related Material
+
+- Does it work with all kind of offline wallets?
+- Generic types currently appear multiple times in the metadata with each instantiation. It could be may be useful to have generic type only once in the metadata and declare the generic parameters at their instantiation. 
```

Comment:
I once tried to do this at the `scale-info` level, see https://github.com/paritytech/scale-info/pull/12, but abandoned it.

It would be considerably easier to do as a second pass over the already generated type information.